### PR TITLE
[stable/artifactory-ha] Added support for ingress in artifactory-ha

### PR DESCRIPTION
**Pull request by jainishshah17**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: stable/artifactory-ha/templates/nginx-deployment.yaml [hunk 1]
```diff
@@ -29,6 +29,17 @@ spec:
       imagePullSecrets:
       - name: {{ .Values.imagePullSecrets }}
     {{- end }}
+      initContainers:
```
Comment (eldada) [inline stable/artifactory-ha/templates/nginx-deployment.yaml]: This initContainer needs to be wrapped with `{{- if .Values.nginx.persistence.enabled }}`

**Review by eldada — CHANGES_REQUESTED** [review]

**Comment by jainishshah17** [comment]
@eldada

**Review by eldada — APPROVED** [review]

File: stable/artifactory-ha/Chart.yaml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 apiVersion: v1
 name: artifactory-ha
 home: https://www.jfrog.com/artifactory/
-version: 0.1.3
+version: 0.1.5
```
Comment (timstoop) [inline stable/artifactory-ha/Chart.yaml]: Not entirely sure why you bumped to 0.1.5, when 0.1.4 is not in use yet. That said, this change warrants a bump of the minor version, as you're adding new features.

File: stable/artifactory-ha/templates/ingress.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      {{ $key }}: {{ $value | quote }}
+    {{- end }}
+spec:
+  backend:
+    serviceName: {{ $serviceName }}
+    servicePort: {{ $servicePort }}
+{{- if .Values.ingress.hosts }}
+  rules:
+  {{- range $host := .Values.ingress.hosts }}
+  - host: {{ $host }}
```
Comment (timstoop) [inline stable/artifactory-ha/templates/ingress.yaml:24]: If you make the host optional (which is allowed by Ingress), it becomes slightly more flexible, I think. Also, you no longer have to provide an example value in values.yaml as set but can comment it out. I think that's cleaner than deploying example values.

File: stable/artifactory/Chart.yaml [hunk 4]
```diff
@@ -1,7 +1,7 @@
 apiVersion: v1
 name: artifactory
 home: https://www.jfrog.com/artifactory/
-version: 7.0.6
+version: 7.0.7
```
Comment (timstoop) [inline stable/artifactory/Chart.yaml:4]: Again, you're adding new features, you should bump the minor version.

**Review by timstoop — CHANGES_REQUESTED** [review]
I noticed you're not adding the Ingress option to the non-ha Chart. Is there a rationale behind it (just curious)?

File: stable/artifactory/Chart.yaml (hunk 4 above)
Comment (jainishshah17) [reply stable/artifactory/Chart.yaml:4]: Ingress support was there. I fixed default backend

File: stable/artifactory-ha/templates/ingress.yaml (hunk 3 above)
Comment (jainishshah17) [reply stable/artifactory-ha/templates/ingress.yaml:24]: done

File: stable/artifactory-ha/Chart.yaml (hunk 2 above)
Comment (jainishshah17) [reply stable/artifactory-ha/Chart.yaml]: fixed

**Comment by carrodher** [comment]
Not sure if, following the policy of one PR per chart, we should split this PRs in two, one for `stable/artifactory-ha` and another one for `stable/artifactory`

**Comment by timstoop** [comment]
It's customary to split them up, easier when PRs are kept small and concise.

**Comment by jainishshah17** [comment]
https://github.com/kubernetes/charts/pull/4373

**Comment by jainishshah17** [comment]
@timstoop I think the test is failing because of wait timing. Can you help with this?

**Comment by prydonius** [comment]
It's very difficult to see what's happening here from the logs, but it _might_ be that the artifactory-ha-member-0 is stuck in a CreateContainerConfigError error:
```
W0401 07:17:26.670] + echo 'INFO: Waiting for pods to enter running state'
W0401 07:17:26.671] + sleep 10
I0401 07:17:26.771] artifac-7425-1-artifactory-ha-member-0              0/1       CreateContainerConfigError   0         26m
I0401 07:17:26.772] artifac-7425-1-artifactory-ha-member-1              0/1       Terminating                  0         22m
I0401 07:17:26.772] artifac-7425-1-artifactory-ha-primary-0             0/1       Terminating                  0         26m
I0401 07:17:26.772] INFO: Waiting for pods to enter running state
I0401 07:17:36.669] ERROR: Some containers failed to reach the ready state
W0401 07:17:36.770] + (( 54 < 54 ))
W0401 07:17:36.770] + (( 1 == 0 ))
W0401 07:17:36.770] + echo 'ERROR: Some containers failed to reach the ready state'
```

From the describe output it looks like it might be failing to mount the ConfigMap:
```
I0401 07:17:38.346]   8m		5m		7	kubelet, gke-jenkins-default-pool-51cafeba-mk3r	spec.containers{artifactory-ha}		Warning		Failed			Error: failed to prepare subPath for volumeMount "binarystore-xml" of container "artifactory-ha"
```

@unguiculus I remember you mentioning some things had changed when mounting ConfigMaps/Secrets in newer K8s versions, could this be related to that?

**Comment by jainishshah17** [comment]
@prydonius If you analyze log closely. `artifactory-ha` was deployed successfully as `artifac-7425-1`. check line `I0401 06:57:27.287` for that.  Then it deletes release line `W0401 06:57:33.361] + helm delete --purge artifac-7425-1`.  At time of Termination it gets stuck with `artifac-7425-1-artifactory-ha-member-0              0/1       CreateContainerConfigError   0         26m`

**Comment by prydonius** [comment]
pinging @unguiculus @mattfarina since you two are more clued in with the CI script. I'm a little bit confused about what's going on here, it looks like we're installing artifactory-ha multiple times even though there's only one values.yaml file to test with. Also, any idea what could cause a pod to go into a `CreateContainerConfigError` loop when terminating?

**Comment by unguiculus** [comment]
First of all, there's two charts in the PR. In general, we want separate PRs per chart. This is why we see to installs in the PR. My I ask you to start off and create separate PRs? Also, you always seem to continue on the same branch which might be outdated. This PR has 64 commits! I'd like to ask you to always start of from a new branch off of the latest master. This ensure the PR runs with the latest CI scripts and will help use trace down issues. Thanks.

**Comment by jainishshah17** [comment]
@unguiculus  I have created new separate PR for both charts.  
https://github.com/kubernetes/charts/pull/4778
https://github.com/kubernetes/charts/pull/4777

**Pull request closed without merging**